### PR TITLE
Feature/error messages

### PR DESCRIPTION
**Pull request by brettwold**

### CODE REVIEW

File: flow-base-api/src/main/java/com/aevi/sdk/flow/constants/ErrorConstants.java [hunk 1]
```diff
@@ -15,19 +15,17 @@
...
+    String PROCESSING_SERVICE_NOT_INSTALLED = "notInstalled";
+    String PROCESSING_SERVICE_BUSY = "busy";
+    String PROCESSING_SERVICE_CANCEL_FAILED = "cancelFailed";
+    String PROCESSING_SERVICE_RESUME_FAILED = "resumeFailed";
     String UNSUPPORTED_OPERATION = "unsupportedOperation";
+    String INVALID_MESSAGE_TYPE = "invalidMessageType";
     String INVALID_REQUEST = "invalidRequest";
-    String UNEXPECTED_ERROR = "unexpectedError";
+    String FLOW_SERVICE_ERROR = "flowServiceError";
+    String FLOW_SERVICE_ACTIVITY_ERROR = "flowServiceActivityError";
```
[the remark this answers is not in the document]
Comment (brettwold) [reply flow-base-api/src/main/java/com/aevi/sdk/flow/constants/ErrorConstants.java]: Forgot :)

File: flow-base-api/src/main/java/com/aevi/sdk/flow/model/FlowException.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    private String errorCode;
+    private String errorMessage;
+
+    public FlowException(String errorCode, String message) {
+        this.errorMessage = message;
+        this.errorCode = errorCode;
+    }
+
+    public String getErrorCode() {
```
[the remark this answers is not in the document]
Comment (brettwold) [reply flow-base-api/src/main/java/com/aevi/sdk/flow/model/FlowException.java]: Lame

File: payment-initiation-api/src/main/java/com/aevi/sdk/pos/flow/PaymentClient.java [hunk 3]
```diff
@@ -48,6 +48,10 @@
      *
      * Returns a single that emits a {@link Response} after processing has completed.
      *
+     * If a fatal error occurs during the flow then a {@link com.aevi.sdk.flow.model.FlowException} will be delivered to the `onError` handler. This
```
[the remark this answers is not in the document]
Comment (brettwold) [reply payment-initiation-api/src/main/java/com/aevi/sdk/pos/flow/PaymentClient.java]: noice... noise :)

File: payment-flow-service-api/src/main/java/com/aevi/sdk/pos/flow/service/BasePaymentFlowService.java [hunk 4]
```diff
@@ -39,42 +40,56 @@ protected void processRequest(@NonNull ClientCommunicator clientCommunicator, @N
...
+                        break;
+
+                }
             }
+        } catch (StageNotImplemented e) {
+            returnStageNotImplemented(clientCommunicator, e.stage);
+        }
+    }
+
+    private class StageNotImplemented extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (brettwold) [reply payment-flow-service-api/src/main/java/com/aevi/sdk/pos/flow/service/BasePaymentFlowService.java]: Well its only internal but if it makes you feel better :)

File: flow-base-api/src/main/java/com/aevi/sdk/flow/BaseApiClient.java [hunk 5]
```diff
@@ -158,6 +179,14 @@ protected ChannelClient getMessengerClient(ComponentName componentName) {
         }
     }
 
+    protected Throwable createFlowException(Throwable throwable) {
+        if (throwable instanceof MessageException) {
+            MessageException e = (MessageException) throwable;
+            return new FlowException(e.getCode(), e.getMessage());
+        }
+        return throwable;
```
Comment (andreas-borglin) [inline flow-base-api/src/main/java/com/aevi/sdk/flow/BaseApiClient.java]: err sorry - late comment. So should we not map any non-FlowException ones here? Like `new FlowException(UNKNOWN_ERROR, throwable.getMessage)` but maybe a new constructor and pass in throwable as the "cause"?

**Pull request merged**